src/combinatorics/hall.lean

```diff
@@ -0,0 +1,341 @@
+  { rw [fintype.card_of_subtype A (λ x, iff.rfl), ← nat.lt_succ_iff],
+    have : A ⊂ univ := ⟨subset_univ _, λ t, hnA (le_antisymm (subset_univ _) t)⟩,
+    apply lt_of_lt_of_le (card_lt_card this) hn },
+  have h4 : (∀ (A' : finset α'), A'.card ≤ (image_rel r' A').card),
+  { intro A',
+    have h₁ := hr (A'.image subtype.val),
+    have h₂ : image_rel r (image subtype.val A') ⊆ image subtype.val (image_rel r' A'),
+    { intro t,
+      simp only [image_rel, mem_filter, mem_image, exists_prop, mem_univ, true_and, subtype.exists,
+        and_imp, exists_eq_right, exists_imp_distrib, exists_and_distrib_right],
+      intros a hA hA' rat,
+      refine ⟨_, _, _, hA', rat⟩,
+      simp only [image_rel, mem_filter, mem_univ, true_and, exists_prop],
+      refine ⟨_, hA, rat⟩ },
+    rw card_image_of_injective _ subtype.val_injective at h₁,
+    apply h₁.trans ((card_le_of_subset h₂).trans _),
+    rw card_image_of_injective _ subtype.val_injective },
+  have h' := ih r' h3 h4,
+  rcases h' with ⟨f', hf', hAf'⟩,
+  let α'' := {a'' : α // a'' ∉ A},
+  let β'' := {b'' : β // b'' ∉ image_rel r A},
+  let r'' : α'' → β'' → Prop := λ a'' b'', r a'' b'',
+  have h5 : fintype.card α'' ≤ n,
+  { have : ¬univ ⊆ Aᶜ,
+    { intro t,
+      rcases hA with ⟨a, ha⟩,
+      simpa [ha] using t (mem_univ a) },
+    have : Aᶜ ⊂ univ := ⟨subset_univ _, this⟩,
+    rw [fintype.card_of_subtype Aᶜ, ← nat.lt_succ_iff],
+    { apply lt_of_lt_of_le (card_lt_card this) hn },
+    { simp } },
+  have h6 : (∀ (B : finset α''), B.card ≤ (image_rel r'' B).card),
+  { intro B,
+    have : (A ∪ B.image subtype.val).card - A.card = B.card,
+    { rw [card_disjoint_union, nat.add_sub_cancel_left,
+        card_image_of_injective _ (subtype.val_injective)],
+      rw disjoint_left,
+      simp only [not_exists, mem_image, exists_prop, exists_and_distrib_right, exists_eq_right,
+        subtype.exists, subtype.coe_mk],
+      intros a hA hA',
+      apply (hA' hA).elim },
+    rw ← this,
+    rw huA,
+    apply (nat.sub_le_sub_right (hr _) _).trans _,
+    rw ← card_sdiff,
+    { have :
+        image_rel r (A ∪ B.image subtype.val) \ image_rel r A ⊆ (image_rel r'' B).image subtype.val,
+      { intros t,
+        simp only [image_rel, mem_filter, mem_sdiff, mem_univ, true_and, exists_prop, not_exists,
+          mem_image, mem_union, and_imp, not_and, exists_and_distrib_right, exists_eq_right,
+          subtype.exists, subtype.coe_mk, exists_imp_distrib],
+        rintro a (ha | ⟨ha, hB⟩) rat hA,
+        { exfalso,
+          apply hA a ha rat },
+        { exact ⟨by simpa [image_rel] using hA, _, ha, hB, rat⟩ } },
+      apply (card_le_of_subset this).trans _,
+      rw card_image_of_injective _ subtype.val_injective },
+    { apply image_rel_mono,
+      apply subset_union_left } },
+  have h'' := ih r'' h5 h6,
+  rcases h'' with ⟨f'', hf'', hAf''⟩,
+  refine ⟨λ x, if h : x ∈ A then f' ⟨x, h⟩ else f'' ⟨x, h⟩, _, _⟩,
+  { rintro x₁ x₂ (h : dite _ _ _ = dite _ _ _),
```

Comment:
I think we talked about this before, but it might be nice to have a lemma saying a function defined by `dite` is injective if the two parts are injective and the images are disjoint